### PR TITLE
feat(security): Remove Vault TLS

### PR DESCRIPTION
**Pull request by jim-wang-intel**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/master/.github/Contributing.md.

## What is the current behavior?

Currently Vault is on TLS.

## What is the new behavior?
Based on the [ADR of remove internal TLS in-cluster communication](https://github.com/edgexfoundry/edgex-docs/pull/287), Vault now is running on Non-TLS, which affects all the configuration toml on all services in which PROTOCOL changes to "http" instead of "https"

## Does this PR introduce a breaking change?

- [ ] Yes
- [X ] No

## New Imports

- [ ] Yes
- [X ] No

## Specific Instructions
Are there any specific instructions or things that should be known prior to reviewing?

### CODE REVIEW

**Review by bnevis-i — CHANGES_REQUESTED** [review]
ServerName should also be set to ""

**Comment by jim-wang-intel** [comment]
> ServerName should also be set to ""

Good catch, will do.  Also plan to do some more renaming of TLS related properties to make it more consistent. 
I just wanted to throw a draft PR out there first for review since the number of lines change would be a bit of big.

File: internal/security/proxy/init.go [hunk 1]
```diff
@@ -74,11 +75,21 @@ func (b *Bootstrap) BootstrapHandler(_ context.Context, _ *sync.WaitGroup, _ sta
 	lc := bootstrapContainer.LoggingClientFrom(dic.Get)
 	configuration := container.ConfigurationFrom(dic.Get)
 
-	req := NewRequestor(
-		b.insecureSkipVerify,
-		configuration.Writable.RequestTimeout,
-		configuration.SecretService.CACertPath,
-		lc)
+	var req internal.HttpCaller
+	if configuration.SecretService.CACertPath != "" {
```
Comment (lenny-intel) [inline internal/security/proxy/init.go:79]: I prefer `len(configuration.SecretService.CACertPath) > 0`

File: internal/security/proxy/init.go [hunk 2]
```diff
@@ -91,15 +102,16 @@ func (b *Bootstrap) BootstrapHandler(_ context.Context, _ *sync.WaitGroup, _ sta
...
+		// var cp bootstrapConfig.CertKeyPair
 
-		cp, err := certificateProvider.GetCertificateKeyPair("kong-tls")
-		if err != nil {
-			b.haltIfError(lc, err)
-		}
+		// cp, err := certificateProvider.GetCertificateKeyPair("kong-tls")
+		// if err != nil {
+		// 	b.haltIfError(lc, err)
+		// }
```
Comment (lenny-intel) [inline internal/security/proxy/init.go:112]: Remove commented out code. I can see what was there previously in this tool.

File: internal/security/proxy/service.go [hunk 3]
```diff
@@ -128,19 +129,20 @@ func (s *Service) ResetProxy() error {
...
+	// 	switch postErr.reason {
+	// 	case CertExisting:
+	// 		s.loggingClient.Info("skipping as the initialization has been done successfully")
+	// 		return nil
+	// 	case InternalError:
+	// 		return errors.New(postErr.Error())
+	// 	default:
+	// 		return errors.New(postErr.Error())
+	// 	}
+	// }
```
Comment (lenny-intel) [inline internal/security/proxy/service.go:145]: Remove commented out code

File: internal/security/secretstore/init.go [hunk 4]
```diff
@@ -363,35 +363,40 @@ func (b *Bootstrap) BootstrapHandler(ctx context.Context, _ *sync.WaitGroup, _ s
...
 	}
 
-	cert := NewCerts(req, configuration.SecretService.CertPath, rootToken, configuration.SecretService.GetSecretSvcBaseURL(), lc)
-	existing, err := cert.AlreadyinStore()
-	if err != nil {
-		lc.Error(err.Error())
-		os.Exit(1)
-	}
+	// With Vault moving to non-TLS mode, the cert related logic is obsolete
+	if configuration.SecretService.CertPath != "" {
```
Comment (lenny-intel) [inline internal/security/secretstore/init.go:367]: Are we going to still support TLS mode if configured or simplify the code to No TLS. @bnevis-i  thoughts?
If we are never going to use TLS, then we should remove the code totally.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: cmd/core-command/res/configuration.toml [hunk 5]
```diff
@@ -48,11 +48,11 @@ Host = 'localhost'
 Port = 8200
 # Use the core-meta data secrets due to core-command using core-meta-data's database for persistance.
 Path = '/v1/secret/edgex/metadata/'
-Protocol = 'https'
-RootCaCertPath = '/vault/config/pki/EdgeXFoundryCA/EdgeXFoundryCA.pem'
+Protocol = 'http'
+RootCaCertPath = ''
```
[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/core-command/res/configuration.toml:52]: My understanding is we need to leave it or we cant override with env vars.

File: internal/security/secretstore/init.go (hunk 4 above)
Comment (bnevis-i) [reply internal/security/secretstore/init.go:367]: Given the subject of distributing microservices across the wire keeps coming up, I am hesitant to go all in.

File: cmd/core-command/res/configuration.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (lenny-intel) [reply cmd/core-command/res/configuration.toml:52]: @bnevis-i,  Only if we want to support TLS option re-enabled via overrides?

File: cmd/security-secretstore-setup/res/configuration.toml [hunk 6]
```diff
@@ -24,13 +25,13 @@ EnableRemote = false
...
 Server = "edgex-vault"
 Port = 8200
-CertPath = "v1/secret/edgex/edgex-security-proxy-setup/kong-tls"
-CaFilePath = "/tmp/edgex/secrets/ca/ca.pem"
-CertFilePath = "/tmp/edgex/secrets/edgex-kong/server.crt"
-KeyFilePath = "/tmp/edgex/secrets/edgex-kong/server.key"
+CertPath = ""      # v1/secret/edgex/edgex-security-proxy-setup/kong-tls if BYO
+CaFilePath = ""
+CertFilePath = ""
+KeyFilePath = ""
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply cmd/security-secretstore-setup/res/configuration.toml:34]: Sounds like for this pass, we will keep these.

File: cmd/security-secretstore-setup/res-file-token-provider/configuration.toml [hunk 7]
```diff
@@ -1,9 +1,9 @@
 [SecretService]
-Scheme = "https"
+Scheme = "http"
 Server = "edgex-vault"
 ServerName = "edgex-vault"
 Port = 8200
-CaFilePath = '/tmp/edgex/secrets/ca/ca.pem'
+CaFilePath = ''
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply cmd/security-secretstore-setup/res-file-token-provider/configuration.toml:6]: Keeping for now.

File: cmd/core-command/res/configuration.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (lenny-intel) [reply cmd/core-command/res/configuration.toml:52]: Keeping for now.

**Comment by lenny-intel** [comment]
Note that we will need to clean-up a bunch of existing overrides in the compose files. Please create an issue on developer-scripts.

File: cmd/security-secretstore-setup/res-file-token-provider/configuration.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply cmd/security-secretstore-setup/res-file-token-provider/configuration.toml:6]: ok, keeping it

File: internal/security/fileprovider/init.go [hunk 8]
```diff
@@ -70,7 +70,7 @@ func (b *Bootstrap) BootstrapHandler(_ context.Context, _ *sync.WaitGroup, _ sta
 		lc.Info("bypassing certificate verification for secret store connection")
 		req = secretstoreclient.NewRequestor(lc).Insecure()
 	}
-	vaultScheme := cfg.SecretService.Scheme
+	vaultScheme := cfg.SecretService.Protocol
```
Comment (lenny-intel) [inline internal/security/fileprovider/init.go:73]: For consistency can we change `vaultScheme` to `vaultProtocol`?

File: internal/security/proxy/config/config.go [hunk 9]
```diff
@@ -87,7 +88,7 @@ type SecretServiceInfo struct {
 
 func (s SecretServiceInfo) GetSecretSvcBaseURL() string {
 	url := &url.URL{
-		Scheme: "https",
+		Scheme: s.Protocol,
```
Comment (lenny-intel) [inline internal/security/proxy/config/config.go:91]: Came we simplify and remove using URL and just use Sprintf. URL is only provider the formatting we can do with Sprintf
```
return fmt.Sprintf("%s://%s:%d", s.Protocol, s.Server, s.Port)
```

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: internal/security/fileprovider/init.go (hunk 8 above)
Comment (jim-wang-intel) [reply internal/security/fileprovider/init.go:73]: yes, that makes sense

File: internal/security/proxy/config/config.go (hunk 9 above)
Comment (jim-wang-intel) [reply internal/security/proxy/config/config.go:91]: sounds good to me

**Review by lenny-intel — APPROVED** [review]
LGTM

**Comment by lenny-intel** [comment]
@jim-wang-intel , I moved this out of draft now that the ADR was approved.

File: internal/security/secretstore/init.go [hunk 10]
```diff
@@ -363,35 +363,40 @@ func (b *Bootstrap) BootstrapHandler(ctx context.Context, _ *sync.WaitGroup, _ s
 		os.Exit(1)
 	}
 
-	cert := NewCerts(req, configuration.SecretService.CertPath, rootToken, configuration.SecretService.GetSecretSvcBaseURL(), lc)
-	existing, err := cert.AlreadyinStore()
-	if err != nil {
-		lc.Error(err.Error())
-		os.Exit(1)
-	}
+	// With Vault moving to non-TLS mode, if cert is not present then no need to upload
```
Comment (bnevis-i) [inline internal/security/secretstore/init.go:366]: Should be more clear in this comment.  The kong token has nothing to do with Vault running in TLS mode or not.
What we are doing is making the kong TLS certificate upload optional as an intermediate step to removing it altogether.

**Review by bnevis-i — COMMENTED** [review]
Reviewed.

**Review by bnevis-i — DISMISSED** [review]

**Review by lenny-intel — DISMISSED** [review]
LGTM

**Review by lenny-intel — APPROVED** [review]

**Review by lenny-intel — APPROVED** [review]

File: internal/security/secretstore/init.go [hunk 11]
```diff
@@ -340,35 +340,40 @@ func (b *Bootstrap) BootstrapHandler(ctx context.Context, _ *sync.WaitGroup, _ s
 		os.Exit(1)
 	}
 
-	cert := NewCerts(req, configuration.SecretService.CertPath, rootToken, configuration.SecretService.GetSecretSvcBaseURL(), lc)
-	existing, err := cert.AlreadyinStore()
-	if err != nil {
-		lc.Error(err.Error())
-		os.Exit(1)
-	}
+	// Uploading Kong's TLS certificate is optional if cert's path is not present
```
Comment (tonyespy) [inline internal/security/secretstore/init.go:343]: See previous comment about Kong cert

File: internal/security/secretstore/password_test.go [hunk 12]
```diff
@@ -81,9 +81,9 @@ func TestRetrieveCred(t *testing.T) {
 
 	configuration := &config.ConfigurationStruct{
 		SecretService: secretstoreclient.SecretServiceInfo{
-			Server: parsed.Hostname(),
-			Port:   port,
-			Scheme: "https",
+			Server:   parsed.Hostname(),
+			Port:     port,
+			Protocol: "https",
```
Comment (tonyespy) [inline internal/security/secretstore/password_test.go:86]: Shouldn't we use "http" here?

**Review by tonyespy — CHANGES_REQUESTED** [review]
Please see comments inline.

Also shouldn't your PR description mention that this is breaking change (`SecretService.Scheme` --> `SecretService.Protocol`), but that it's OK due to Hanoi being a 2.0 release? I wonder if we should start labeling all PRs with breaking configuration key changes so we'll have an inventory of changes that we can use for upgrade logic?

File: cmd/security-bootstrap-redis/res/configuration.toml [hunk 13]
```diff
@@ -22,13 +22,13 @@ File = ''
 
 [SecretStore]
 Host = 'localhost'          ## Override in environment variables, if necessary
-ServerName = 'edgex-vault'  ## Needed for TLS CN
+ServerName = ''             ## Needed for TLS CN
```
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply cmd/security-bootstrap-redis/res/configuration.toml:25]: yes, good catch

File: cmd/core-command/res/configuration.toml [hunk 14]
```diff
@@ -48,11 +48,11 @@ Host = 'localhost'
 Port = 8200
 # Use the core-meta data secrets due to core-command using core-meta-data's database for persistance.
 Path = '/v1/secret/edgex/metadata/'
-Protocol = 'https'
-RootCaCertPath = '/vault/config/pki/EdgeXFoundryCA/EdgeXFoundryCA.pem'
-ServerName = 'localhost'
+Protocol = 'http'
+RootCaCertPath = ''
+ServerName = ''
```
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply cmd/core-command/res/configuration.toml:53]: @tonyespy this `ServerName` is only being used when there is TLS communication to Vault as of today.  Since we are removing Vault TLS, so we can blank all them out.

File: cmd/security-secretstore-setup/res/configuration.toml [hunk 15]
```diff
@@ -24,13 +25,13 @@ EnableRemote = false
 File = ""
 
 [SecretService]
-Scheme = "https"
+Protocol = "http"
 Server = "edgex-vault"
 Port = 8200
-CertPath = "v1/secret/edgex/edgex-security-proxy-setup/kong-tls"
```
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply cmd/security-secretstore-setup/res/configuration.toml:30]: yea, can leave it as blank-out

File: internal/security/secretstore/init.go (hunk 11 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/init.go:343]: We are making this optional and deprecated eventually since there is `secrets-config proxy tls` to accomplish this.  @bnevis-i , please chime-in for more details if I am incorrect.

File: internal/security/secretstore/password_test.go (hunk 12 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/password_test.go:86]: Negative, this is for the already existing test case.

File: internal/security/secretstore/init.go (hunk 11 above)
Comment (bnevis-i) [reply internal/security/secretstore/init.go:343]: There are two imminent PR's regarding removal of TLS that will follow this one.  The first is (1) don't generate TLS certificates at all (there won't be a cert to upload), (2) use "secrets-config proxy tls" to upload your BYO cert (just need to port one function call out of proxy-setup to add the missing functionality), (3) delete this code for uploading TLS certs into Vault entirely (after previous two done).

File: cmd/security-secretstore-setup/res/configuration.toml (hunk 15 above)
[the remark this answers is not in the document]
Comment (lenny-intel) [reply cmd/security-secretstore-setup/res/configuration.toml:30]: @jim-wang-intel , question is why don't we need this since it is the TLS cert for Kong, not for Vault?

**Comment by tonyespy** [comment]
FYI, pretty much confirmed my theory about `SecretStore.ServerName` and the snap. If I build the snap from your branch, it fails to install due to misconfiguration of `SecretStore`. I'm working on a patch to snapcraft.yaml for you.

Also FYI, your branch has conflicts which also need resolving.

File: cmd/security-secretstore-setup/res/configuration.toml (hunk 15 above)
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply cmd/security-secretstore-setup/res/configuration.toml:30]: we don't need this since Kong will generate cert itself and we are in the stage of removing TLS communication internally.

**Review by bnevis-i — DISMISSED** [review]
Aside from other problems that @tonyespy brought up, the Docker smoke tests seem to run fine.  Note that need to change VAULT_ADDR to http:// in docker-compose (nonblocking, only affects interactive use of Vault)

File: cmd/core-command/res/configuration.toml (hunk 14 above)
[the remark this answers is not in the document]
Comment (tonyespy) [reply cmd/core-command/res/configuration.toml:53]: OK, thanks for clarifying.

File: internal/security/secretstore/init.go (hunk 11 above)
Comment (tonyespy) [reply internal/security/secretstore/init.go:343]: So we're completely removing the auto-generated EdgeXCA signed TLS certificate for Kong too?

If so, IMHOP it would've been better to separate out those changes into a different PR.

File: cmd/security-secretstore-setup/res/configuration.toml (hunk 15 above)
[the remark this answers is not in the document]
Comment (tonyespy) [reply cmd/security-secretstore-setup/res/configuration.toml:25]: @bnevis-i care to comment on this? This PR is about disabling TLS for Vault, yet we're also removing the upload of the Kong EdgeXCA signed TLS certificate to Vault too. If this cert isn't uploaded to Vault does this trigger the automatic selfsigned TLS cert generation logic in Kong or are there are additional changes that need to be made to enable that to work? Also if Kong generates its own TLS cert, how does an external client validate this cert?

File: cmd/security-secrets-setup/start_vault.sh [hunk 16]
```diff
@@ -24,21 +24,16 @@ VAULT_TLS_PATH=${VAULT_TLS_PATH:-/tmp/edgex/secrets/edgex-vault}
 DEFAULT_VAULT_LOCAL_CONFIG='
 listener "tcp" { 
               address = "edgex-vault:8200" 
-              tls_disable = "0" 
+              tls_disable = "1"
```
Comment (tonyespy) [inline cmd/security-secrets-setup/start_vault.sh:27]: Should we also get rid of the `VAULT_TLS_PATH` logic in this script, and in particular the `SENTILNEL_FILE` loop?

**Review by tonyespy — COMMENTED** [review]
See inline comments... (down to two).

Also I'm attaching two patch files for you, one to update `snapcraft.yaml`, and the other to update the vault config file used in the snap. Could you please apply both to your PR? I've confirmed that with both changes the snap builds and installs.

[0001-fix-snap-disable-TLS-in-vault-config-file.patch.txt](https://github.com/edgexfoundry/edgex-go/files/5639724/0001-fix-snap-disable-TLS-in-vault-config-file.patch.txt)

[0002-fix-snap-remove-vault-TLS-config-overrides.patch.txt](https://github.com/edgexfoundry/edgex-go/files/5639726/0002-fix-snap-remove-vault-TLS-config-overrides.patch.txt)

**Comment by tonyespy** [comment]
OK for some reason GitHub ignored by comment on cmd/security-secretstore-setup/res/configuration.toml because Jim marked the conversation as resolved, and somehow it wouldn't let me re-open it.  Anyways, the specific comment was:

@bnevis-i care to comment on this? This PR is about disabling TLS for Vault, yet we're also removing the upload of the Kong EdgeXCA signed TLS certificate to Vault too. If this cert isn't uploaded to Vault does this trigger the automatic selfsigned TLS cert generation logic in Kong or are there are additional changes that need to be made to enable that to work? Also if Kong generates its own TLS cert, how does an external client validate this cert?

**Comment by bnevis-i** [comment]
> @bnevis-i care to comment on this? This PR is about disabling TLS for Vault, yet we're also removing the upload of the Kong EdgeXCA signed TLS certificate to Vault too. If this cert isn't uploaded to Vault does this trigger the automatic selfsigned TLS cert generation logic in Kong or are there are additional changes that need to be made to enable that to work? Also if Kong generates its own TLS cert, how does an external client validate this cert?

I spun up the framework earlier this afternoon.  Kong serves its default self-signed TLS certificate.  This is no worse and no than today's situation where we create a CA and leaf certificate out of thin air that is also not trusted.

File: cmd/security-secrets-setup/start_vault.sh (hunk 16 above)
Comment (jim-wang-intel) [reply cmd/security-secrets-setup/start_vault.sh:27]: Tony @tonyespy, at this stage, we still need that since the `secrets-setup` has not been removed yet and thus docker consul health checkers still checks that I guess.  Removing secrets-setup is another big step on snap as well.

File: internal/security/secretstore/init.go (hunk 11 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/init.go:343]: I think we are saying, at this stage, the `secrets-setup` still runs and thus generated certs for both but we are turning some default behavior to non-TLS as Vault has changed to non-TLS by default (but it does not mean that you cannot overwrite it in the docker-compose file with using environment variable `LOCAL_VAULT_CONFIG` to turn TLS back on before compose up)  so that is why this turns that into optional for now.

**Comment by jim-wang-intel** [comment]
@tonyespy I think I addressed your comments, please let me know if it works for you.  Thanks!

File: internal/security/secretstore/init.go (hunk 11 above)
Comment (bnevis-i) [reply internal/security/secretstore/init.go:343]: @jim-wang-intel 

If we backed out this change to set these keys to the empty string
```
CertPath = "v1/secret/edgex/edgex-security-proxy-setup/kong-tls"
CaFilePath = "/tmp/edgex/secrets/ca/ca.pem"
CertFilePath = "/tmp/edgex/secrets/edgex-kong/server.crt"
KeyFilePath = "/tmp/edgex/secrets/edgex-kong/server.key"
```

would that re-enable the Kong TLS from the internal CA?

We could then put this change into the PR that deletes security-secrets-setup.

Comment (jim-wang-intel) [reply internal/security/secretstore/init.go:343]: That PR deleting security-secrets-setup is getting messier, I wouldn't want to add too much change in one PR.

Comment (bnevis-i) [reply internal/security/secretstore/init.go:343]: Maybe delete TLS configuration in one PR, and delete code in another?  I am trying to offer a solution that will meet both your and Tony's needs.

Comment (jim-wang-intel) [reply internal/security/secretstore/init.go]: @bnevis-i  Done.

**Comment by tonyespy** [comment]
> Kong serves its default self-signed TLS certificate.

OK, I would've rather seen this in a separate PR as the title of this PR is "Remove Vault TLS", but I've made my arguments and apparently they weren't convincing enough, so I'll capitulate and approve "as is".

File: cmd/security-proxy-setup/res/configuration.toml [hunk 17]
```diff
@@ -17,7 +17,7 @@
 # This is a TOML config file for edgexsecurity service.
 
 [Writable]
-LogLevel = "ERROR"
+LogLevel = "INFO"
```
Comment (tonyespy) [inline cmd/security-proxy-setup/res/configuration.toml:20]: Why was this changed?

**Review by tonyespy — COMMENTED** [review]
One quick question (see inline) about whether a changed `LogLevel` was intentional?

**Comment by tonyespy** [comment]
> > Kong serves its default self-signed TLS certificate.
> 
> OK, I would've rather seen this in a separate PR as the title of this PR is "Remove Vault TLS", but I've made my arguments and apparently they weren't convincing enough, so I'll capitulate and approve "as is".

Darn, it looks my advice was taken and the change backed out. Much appreciated @jim-wang-intel. I'm testing the snap now, and other than my question about the LogLevel change, we should be able to merge this shortly.

File: cmd/security-proxy-setup/res/configuration.toml (hunk 17 above)
Comment (jim-wang-intel) [reply cmd/security-proxy-setup/res/configuration.toml:20]: Good catch. It came through the merged conflicts with master branch yesterday.  Eventually we want it to be INFO to start with based on another issue by Cloud.

**Comment by jim-wang-intel** [comment]
> > > Kong serves its default self-signed TLS certificate.
> > 
> > 
> > OK, I would've rather seen this in a separate PR as the title of this PR is "Remove Vault TLS", but I've made my arguments and apparently they weren't convincing enough, so I'll capitulate and approve "as is".
> 
> Darn, it looks my advice was taken and the change backed out. Much appreciated @jim-wang-intel. I'm testing the snap now, and other than my question about the LogLevel change, we should be able to merge this shortly.

Thanks, Tony @tonyespy , appreciate your help on testing!

File: cmd/security-proxy-setup/res/configuration.toml (hunk 17 above)
Comment (jim-wang-intel) [reply cmd/security-proxy-setup/res/configuration.toml]: changed has been reverted

**Review by tonyespy — APPROVED** [review]
Thanks for the change @jim-wang-intel!

LGTM
